runtime/src/main/java/org/corfudb/runtime/collections/CorfuTable.java

```diff
@@ -197,65 +197,85 @@ public int hashCode() {
      * @param <V>                   Value type
      * @return                      A type token to pass to the builder.
      */
-    static <K, V>
-    TypeToken<CorfuTable<K, V>>
-        getTableType() {
-            return new TypeToken<CorfuTable<K, V>>() {};
+    static <K, V> TypeToken<CorfuTable<K, V>> getTableType() {
+        return new TypeToken<CorfuTable<K, V>>() {};
     }
 
+    // The "main" map which contains the primary key-value mappings.
+    private final StreamingMap<K,V> mainMap;
+    private final Set<Index<K, V, ? extends Comparable>> indexSpec;
+    private final Map<String, Map<Comparable, Map<K, V>>> secondaryIndexes;
+    private final CorfuTable<K, V> optimisticTable;
+    private final VersionPolicy versionPolicy;
+
+    public CorfuTable(StreamingMap<K,V> mainMap,
```

Comment:
Why overload the CorfuTable definition and not create a new type?

If this is being used for the large table use cases, allowing secondary indexes (using this implementation) can become an issue. I think not allowing secondary indexes is better than hoping that it won't be used.

Reply:
No good reason -- mostly to avoid code duplication.

Reply:
I think we should split it, the way its structured right now is error-prone. Actually, there's a leak. If you look at `TableRegistry::openTable` an instance of `PersistedStreamingMap` is passed, which itself isn't SMR, but passed to CorfuTable to be used during construction. This means that if the outer object (CorfuTable) needs to be reconstructed (i.e. TrimmedException/NoRollBackException) it will use state from the old instance. In other words, the same RocksDB instances can be used multiple times for different CorfuTable instances. Imagine the following scenario:
1. CorfuTable is created
2. CorfuTable is access at irregular intervals which results in intermittent TrimmedExceptions (handled internally)
3. Every time a trim exception is encountered, the VLO will invoke the CorfuTable constructor (to create a new object)
4. The object from step 3 will have state from the previous instance (since the PersistedStreamingMap is not constructed as a part of CorfuTable, but passed as a reference)

In essence, the underlying resources will never be closed/cleaned and therefore can grow overtime. 
Decoupling those two doesn't make a lot of sense (to me at least), I think Its possible to do code sharing in a different way.

Reply:
There are two concerns here:

1) Resetting the SMR object when we hit some sort of an unrecoverable error. The bug you are referring to can be addressed by passing a Supplier<T> instead of T.

2) Figuring out how to split memory and disk implementations into two different classes.

I agree that we should do (2), but my suggestion is to do it in a subsequent patch, mostly because the blast radius is pretty large, and figuring out how to do it cleanly is going to take some time.

Reply:
Fixed. `CorfuTable` now takes `Supplier<T>`.